lib/Yield/Support/SugiyamaStyleGraphLayout/GraphPreparation.cpp

```diff
@@ -213,21 +213,26 @@ void partition(const std::vector<EdgeType> &Edges,
   }
 }
 
-/// Breaks long edges into partitions by introducing new internal nodes.
-template<RankingStrategy Strategy>
-RankContainer
-partitionLongEdges(InternalGraph &Graph, NodeClassifier<Strategy> &Classifier) {
-  // To simplify the ranking algorithms, if there's more than one entry point,
-  // an artifitial entry node is added. This new node has a single edge per
-  // real entry point.
+/// To simplify the ranking algorithms, if there's more than one entry point,
+/// an artifitial entry node is added. This new node has a single edge per
+/// real entry point.
+static void
+ensureSingleEntry(InternalGraph &Graph, RankContainer *MaybeRanks = nullptr) {
   llvm::SmallVector<NodeView, 4> EntryNodes;
   for (auto *Node : Graph.nodes())
     if (!Node->hasPredecessors())
       EntryNodes.emplace_back(Node);
   revng_assert(!EntryNodes.empty());
 
-  revng_assert(Graph.getEntryNode() == nullptr);
   if (EntryNodes.size() > 1) {
+    if (Graph.getEntryNode() != nullptr) {
+      if (Graph.getEntryNode()->isVirtual()) {
+        Graph.removeNode(Graph.getEntryNode());
+        if (MaybeRanks != nullptr)
+          MaybeRanks->erase(Graph.getEntryNode());
+      }
+    }
```

Comment:
This (https://github.com/revng/revng/pull/191/commits/50b7a5bac6c740f61d966bbffff91317e7c05bdc) should be the solution for the possible problems.

Reply:
Ok for my comment on `ensureSingleEntry`. But I still don't understand very clearly what this piece of code does.

If I read it correctly: whenever we find many entry points, and the graph has an EntryNode which is virtual, that virtual entry node is removed from the graph and from the ranking. Why so?

Also, this reminds me that to my new implementation of `entryPoints` does not use `Graph->EntryPoint` at all, but it should. @Cvelth can you please fix it? It should be easy, just to a `llvm::post_order_ext(NodeRef(Graph->getEntryNode(), Visited)` before the first iteration on `llvm::nodes` in there, and explicitly add `Graph->getEntryNode()` to `Result`.

Reply:
> But I still don't understand very clearly why this code does this dance.

It's simpler than that, all of the ranking strategies expect the graph to have a single entry point to work correctly, I guess there's an option of reworking them, but I'd rather not go there unless necessary.

> But still, why so?

Virtual "entry point" indicates that it was artificially added there. We don't want to chain those if we can. That's mostly safeguard from future modifications - what if similar function was already invoked on the same graph?

> can you please fix it?

The original requirement set by @aleclearmind was for this function to be graph-agnostic, i.e. it can only rely on the `llvm::GraphTraits`, as such `Graph.getEntryPoint()` is off bounds. I can try to fix something up using `if constexpr(StringSpecializationfOf<GraphType, GenericGraph>)`, but IMO, it's better if we left it as is, let it ignore the "set" entry point. It's not like it modifies graph or anything - it just returns a minimal set of points needed to full traverse the graph.

Reply:
Something like this? https://github.com/revng/revng/pull/191/commits/70adaf68543be1117a3c80c2d546a5c66c9f0b87